k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,383 @@
+ *
+ * @code{.cpp}
+ * _K2CudaDebug(cudaGetLastError(), __FILE__, __LINE__);
+ * @endcode
+ */
+__host__ __device__ __forceinline__ cudaError_t _K2CudaDebug(
+    cudaError_t error,
+    const char *filename,
+    int line) {
+#ifdef K2_MAKE_ERROR_CHECK
+  (void)filename;
+  (void)line;
+  if (cudaSuccess != error) {
+  #if (K2_PTX_ARCH == 0)
+    fprintf(stderr, "CUDA error ID=%d, NAME=%s, [%s, %d]: %s\n",
+            error, cudaGetErrorName(error),
+            filename, line,
+            cudaGetErrorString(error));
+    fflush(stderr);
+  #elif (K2_PTX_ARCH >= 200)
+    printf("CUDA error ID=%d, NAME=%s, "
+           "[block (%d,%d,%d) thread (%d,%d,%d), %s, %d]: %s\n",
+           error, cudaGetErrorName(error),
+           blockIdx.x, blockIdx.y, blockIdx.z,
+           threadIdx.x, threadIdx.y, threadIdx.z,
+           filename, line,
+           cudaGetErrorString(error));
+  #endif
+  }
+#endif
+  return error;
+}
+
+/**
+ * @def K2_CUDA_CHECK_ERROR(cudaError)
+ *
+ * @brief Macro for checking cuda error.
+ *
+ * @details
+ *  If `K2_MAKE_ERROR_CHECK` is defined and e is not cudaSuccess,
+ *  print the error message and exit with the error enum value.
+ *  Otherwise, it does nothing.
+ *
+ * @param[in] e an enum type indicating CUDA errors.
+ *
+ * @remark
+ *  Device cannot terminate/exit the host process. And calling a
+ *  `__host__ function("exit")` from a `__host__ __device__ function`
+ *  is not allowed. Thus, the exit(e) is put here, rather than inside of
+ *  `_K2CudaDebug`.
+ *
+ * @todo
+ *  Should the error code return by this macro?
+ *
+ * @code{.cpp}
+ * K2_CUDA_CHECK_ERROR(cudaGetLastError());
+ * @endcode
+ */
+#ifndef K2_CUDA_CHECK_ERROR
+  #ifdef K2_MAKE_ERROR_CHECK
+    #define K2_CUDA_CHECK_ERROR(e)                                    \
+      if (::k2::_K2CudaDebug((cudaError_t)(e), __FILE__, __LINE__)) { \
+        exit(e);                                                      \
+      }
+  #else
+    #define K2_CUDA_CHECK_ERROR(e) ((void)0)
+  #endif
+#endif
+
+/**
+ * @def K2_CUDA_API_SAFE_CALL(CUDA_RUNTIME_API)
+ *
+ * @brief Macro for checking cuda standard runtime api return status.
+ *
+ * @details
+ *  If api return status is not cudaSuccess, print and exit.
+ *  Thus it's actually could be an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * @note
+ *  All runtime api return an error code. But for an asynchronous api,
+ *  the error code only reports errors that occur on the host
+ *  prior to executing the task, typically related to parameter validation.
+ *  The device kernel error would unfortunately be left for next runtime api
+ *  call. To check the whole process of cuda runtime async api,
+ *  as cuda kernel launches are asynchronous also,
+ *  the same macro for kernel could be used. See
+ *  `K2_CUDA_KERNEL_SAFE_CALL`.
+ *
+ * @code{.cpp}
+ * K2_CUDA_API_SAFE_CALL(CUDA_RUNTIME_API());
+ * @endcode
+ */
+#ifndef K2_CUDA_API_SAFE_CALL
+  #ifdef K2_MAKE_ERROR_CHECK
```

Comment:
I don't understand why `K2_MAKE_ERROR_CHECK` is named this way.  What does `MAKE` signify?  
This only seems to get turned on if NDEBUG is not defined... how is it different from NDEBUG?

Reply:
If NDEBUG, and `K2_MAKE_ERROR_CHECK` is defined manually before include "debug.h". The macros still take effect (except assert that under NDEBUG control). The macro name is bad and could be changed.

Reply:
> What does MAKE signify?

I guess it means `enable`.